File: src/main/java/com/university/services/impl/UserServiceImpl.java
package com.university.services.impl;

import com.university.entities.Department;
import com.university.entities.Position;
import com.university.entities.User;
import com.university.services.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import com.university.repositories.UserRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class UserServiceImpl implements UserService {

    private final UserRepository userRepository;

    @Autowired
    public UserServiceImpl(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public void saveUser(User user, boolean edit) {
        userRepository.saveAndFlush(user);
    }

    @Override
    public void removeUser(int id) {
        userRepository.delete(id);
    }

    @Override
    public User getUserByEmail(String email) {
        return userRepository.findByEmail(email);
    }

    @Override
    public User getUserById(int id) {
        return userRepository.findOne(id);
    }

    @Override
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    @Override
    public List<User> getUsersByFilter(Optional<Department> department, Optional<Position> position) {
        if (department.isPresent() && position.isPresent()) {
            return getUsersByDepartmentAndPosition(department.get(), position.get());
        } else if (department.isPresent()) {
            return getUsersByDepartment(department.get());
        } else if (position.isPresent()) {
            return getUsersByPosition(position.get());
        } else {
            return getAllUsers();
        }
    }

    @Override
    public List<User> getUsersByDepartmentAndPosition(Department department, Position position) {
        return userRepository.findUserByDepartmentAndPosition(department, position);
    }

    @Override
    public List<User> getUsersByDepartment(Department department) {
        return userRepository.findUserByDepartment(department);
    }

    @Override
    public List<User> getUsersByPosition(Position position) {
        return userRepository.findUserByPosition(position);
    }
}

File: src/main/kotlin/com/university/services/impl/UserServiceImpl.kt
package com.university.services.impl

import com.university.entities.Department
import com.university.entities.Position
import com.university.entities.User
import com.university.repositories.UserRepository
import com.university.services.UserService
import org.springframework.stereotype.Service

@Service
class UserServiceImpl (private val userRepository: UserRepository) : UserService {
    override fun save(user: User, edit: Boolean) {
        userRepository.saveAndFlush(user)
    }

    override fun remove(id: Int) {
        userRepository.deleteById(id)
    }

    override fun findByEmail(email: String): User {
        return userRepository.findByEmail(email)
    }

    override fun findById(id: Int): User {
        return userRepository.findById(id).orElseThrow{RuntimeException()}
    }

    override fun getAll(): List<User> {
        return userRepository.findAll()
    }

    override fun findByFilter(department: Department?, position: Position?): List<User> {
        return if (department != null && position != null) {
            findByDepartmentAndPosition(department, position)
        } else if (department != null) {
            findByDepartment(department)
        } else if (position != null) {
            findByPosition(position)
        } else {
            getAll()
        }
    }

    override fun findByDepartmentAndPosition(department: Department, position: Position): List<User> {
        return userRepository.findUserByDepartmentAndPosition(department, position)
    }

    override fun findByDepartment(department: Department): List<User> {
        return userRepository.findUserByDepartment(department)
    }

    override fun findByPosition(position: Position): List<User> {
        return userRepository.findUserByPosition(position)
    }
}
